fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/CascadesPlanner.java

```diff
@@ -502,58 +558,48 @@ public interface Task {
      */
     private class OptimizeGroup implements Task {
         @Nonnull
-        private final PlanContext context;
+        final PlannerPhase plannerPhase;
         @Nonnull
         private final Reference group;
-        @Nonnull
-        private final EvaluationContext evaluationContext;
 
-        public OptimizeGroup(@Nonnull PlanContext context,
-                             @Nonnull Reference group,
-                             @Nonnull final EvaluationContext evaluationContext) {
-            this.context = context;
+        public OptimizeGroup(@Nonnull final PlannerPhase plannerPhase,
+                             @Nonnull final Reference group) {
+            this.plannerPhase = plannerPhase;
             this.group = group;
-            this.evaluationContext = evaluationContext;
+        }
+
+        @Nonnull
+        @Override
+        public PlannerPhase getPlannerPhase() {
+            return plannerPhase;
         }
 
         @Override
         public void execute() {
-            if (group.needsExploration()) {
-                // Explore the group, then come back here to pick an optimal expression.
-                taskStack.push(this);
-                for (RelationalExpression member : group.getMembers()) {
-                    // enqueue explore expression which then in turn enqueues necessary rules for transformations
-                    // and matching
-                    exploreExpressionAndOptimizeInputs(context, group, member, false, evaluationContext);
-                }
-                // the second time around we want to visit the else and prune the plan space
-                group.startExploration();
-            } else {
-                RelationalExpression bestMember = null;
-                for (RelationalExpression member : group.getMembers()) {
-                    if (bestMember == null || new CascadesCostModel(configuration).compare(member, bestMember) < 0) {
-                        if (bestMember != null) {
-                            // best member is being pruned
-                            traversal.removeExpression(group, bestMember);
-                        }
-                        bestMember = member;
-                    } else {
-                        // member is being pruned
-                        traversal.removeExpression(group, member);
+            RelationalExpression bestFinalExpression = null;
```

Comment:
Not sure why this `if(group.needsExploration()) {..} else {...}` was there to begin with. I convinced myself that this code just hid away an implicit `ExploreGroup`. The change together with the additional `ExploreGroup` tasks that gets pushed when a new phase is initiated seems to be identical and the code is much cleaner.